### PR TITLE
ZF3 support

### PR DESCRIPTION
**Pull request by webimpress**
Full ZF3 support. Dropped support PHP 5.5 and ZF2.
It works with https://github.com/doctrine/DoctrineModule/pull/564.

Need a bit more work.
Documentation needs update.

Any suggestions would be appreciated.

(library uses now PSR-4 autoloading, as mentioned in https://github.com/doctrine/DoctrineModule/pull/564 doctrine uses deprecated PSR-0, it could be moved back to PSR-0 later on if needed)

### CODE REVIEW

**Comment by Saeven** [comment]
Quick tag a release! :)  Very exciting stuff!

File: .travis.yml [hunk 1]
```diff
@@ -1,23 +1,66 @@
...
-  - 7
-  - hhvm
+cache:
+  directories:
+    - $HOME/.composer/cache
+    - vendor
+
+env:
+  global:
+    - COMPOSER_ARGS="--no-interaction --ignore-platform-reqs"
```
Comment (xtreamwayz) [inline .travis.yml]: This line is causing the tests to fail. AFAIK this ignores the PHP version as well. That's why it installs PackageVersions which has a PHP 7 dependency. And thus it fails on PHP 5.6 with the latest dependencies.

Comment (webimpress) [inline .travis.yml]: @xtreamwayz thanks! Removed. Now all tests pass.

File: config/module.config.php [hunk 2]
```diff
@@ -79,117 +79,117 @@
...
+        ],
+    ],
+
+    'service_manager' => [
+        'factories' => [
             'Doctrine\ORM\EntityManager' => 'DoctrineORMModule\Service\EntityManagerAliasCompatFactory',
-        ),
-        'invokables' => array(
+        ],
+        'invokables' => [
```
Comment (svycka) [inline config/module.config.php]: maybe remove this?
https://zendframework.github.io/zend-servicemanager/migration/#invokables

File: src/Module.php [hunk 3]
```diff
@@ -58,48 +60,47 @@ function () use ($manager) {
...
     public function getConfig()
     {
-        return include __DIR__ . '/../../config/module.config.php';
+        return include __DIR__ . '/../config/module.config.php';
     }
 
     /**
      * {@inheritDoc}
      */
     public function getControllerConfig()
```
Comment (svycka) [inline src/Module.php]: maybe move this in module.config.php and remove?

**Comment by svycka** [comment]
maybe wrong place to ask but what about plans moving "zendframework/zend-developer-tools", "doctrine/migrations" and "zendframework/zend-hydrator"  into separate projects?

**Comment by Ocramius** [comment]
@webimpress they are separate projects?

**Comment by webimpress** [comment]
@Ocramius @svycka hm... how do you mean? what do you want to do?

**Comment by webimpress** [comment]
rebased after 1.0.0 release

**Comment by svycka** [comment]
@webimpress I mean we agreed to remove hydrator from this module and move hydrator into its own repository also remove zenddevelopertools from required dependancies and maybe create one more repository with the code we have here for it. But maybe this would be to much for this PR.

also maybe we could do the same way zend-stdlib did with zend-hydrator that would allow people migrate when they are ready, but in the mean time they could use ZF3. Just a thought may be not possible.

**Comment by MatthiasKuehneEllerhold** [comment]
The fix for the "zendDeveloperToolsTime" view helper is also available upstream here: https://github.com/zendframework/ZendDeveloperTools/pull/222

**Comment by gianarb** [comment]
@svycka it's a good point but out of this pull request :)

File: src/Module.php (hunk 3 above)
Comment (gianarb) [inline src/Module.php]: We can do that in another PR

File: config/module.config.php (hunk 2 above)
Comment (gianarb) [inline config/module.config.php]: 👍 good catch!

**Comment by Maaslos-UG** [comment]
There is a another Problem with ZF3

Maximum Version of zend-mvc is 2.5, and the actual version is 3.0, doctrine doesn't install via composer: 
 Problem 1
    - Installation request for doctrine/doctrine-orm-module ^1.0 -> satisfiable by doctrine/doctrine-orm-module[1.0.0].
    - doctrine/doctrine-orm-module 1.0.0 requires zendframework/zend-mvc ^2.5.2 -> satisfiable by zendframework/zend-mvc[2.5.2, 2.5.3, 2.6.0, 2.6.1, 2.6.2, 2.6.3, 2.7.0, 2.7.1, 2.7.10, 2.7.2, 2.7.3, 2.7.4, 2.7.5, 2.7.6, 2.7.7, 2.7.8, 2.7.9] but these conflict with your requirements or minimum-stability.

**Comment by localheinz** [comment]
@Maaslos-UG 

This seems like a lot of unrelated changes, how about breaking this apart into multiple PRs?

File: .travis.yml [hunk 4]
```diff
@@ -1,23 +1,74 @@
...
+  - if [[ $DEPS == 'lowest' ]]; then travis_retry composer update --prefer-lowest --prefer-stable $COMPOSER_ARGS ; fi
+  - travis_retry composer install $COMPOSER_ARGS
+  - composer show
 
 script:
-  - ./vendor/bin/phpunit --coverage-clover ./build/clover.xml
-  - if [ $TRAVIS_PHP_VERSION != 'hhvm' ] && [ $TRAVIS_PHP_VERSION != '7' ]; then php build/coverage-checker.php build/clover.xml 70; fi
-  - ./vendor/bin/phpcs --standard=PSR2 ./src/ ./tests/
+  - if [[ $TEST_COVERAGE == 'true' ]]; then composer test-coverage ; else composer test ; fi
+  - if [[ $CHECK_CS == 'true' ]]; then composer cs ; fi
```
Comment (localheinz) [inline .travis.yml]: @Maaslos-UG 

For example, providing composer scripts could be a separate PR, integrated quickly into `master`.

File: README.md [hunk 5]
```diff
@@ -31,30 +31,30 @@ configuration for each of your entities namespaces:
 
 ```php
 <?php
-return array(
-    'doctrine' => array(
-        'driver' => array(
+return [
```
Comment (localheinz) [inline README.md]: @Maaslos-UG 

All of these coding style fixes have nothing to do with making this project compatible with ZF3.

Can we apply coding style fixes in a separate PR, one that addresses coding style fixes only?

File: composer.json [hunk 6]
```diff
@@ -31,40 +31,60 @@
...
+        "zendframework/zend-modulemanager":   "^2.7.2"
     },
     "suggest":           {
         "zendframework/zend-form":            "if you want to use form elements backed by Doctrine",
         "zendframework/zend-developer-tools": "zend-developer-tools if you want to profile operations executed by the ORM during development",
         "doctrine/migrations":                "doctrine migrations if you want to keep your schema definitions versioned"
     },
     "autoload":          {
-        "psr-0": {
+        "psr-4": {
```
Comment (localheinz) [inline composer.json]: @Maaslos-UG 

Switching from PSR-0 to PSR-0 could be addressed in a separate PR, too.

What do you think?

File: src/Form/Annotation/AnnotationBuilder.php [hunk 7]
```diff
@@ -76,31 +77,31 @@ public function getFormSpecification($entity)
         $metadata     = $this->objectManager->getClassMetadata(is_object($entity) ? get_class($entity) : $entity);
         $inputFilter  = $formSpec['input_filter'];
 
-        $formElements = array(
-            'DoctrineModule\Form\Element\ObjectSelect',
-            'DoctrineModule\Form\Element\ObjectMultiCheckbox',
-            'DoctrineModule\Form\Element\ObjectRadio',
-        );
+        $formElements = [
+            Element\ObjectSelect::class,
```
Comment (localheinz) [inline src/Form/Annotation/AnnotationBuilder.php]: @Maaslos-UG 

Using the `class` keyword instead of string literals could be addressed in a separate PR, too.

Personally, I'd always suggest to break large PRs into a number of small PRs, especially if they address entirely separate issues.

I'm pretty sure that 75% of what is addressed in this PR could already have been merged into `master` if broken apart into separate PRs.

What do you think?

Comment (webimpress) [inline src/Form/Annotation/AnnotationBuilder.php]: @localheinz in general you have right, it will be better to have everything in separate PR, but then we will have chain of PRs dependent on each other and if something will be missed is some of the beginning PR then will be very difficult to update all of it.

You can try cherry-pick commits from my PR and create separate PRs but I'm not sure if it is worth.

**Comment by webimpress** [comment]
There is some problem with travis, builds failed for lowest and latests dependencies. Any ideas what is wrong?

**Comment by MatthiasKuehneEllerhold** [comment]
Thats weird:
"Source directory /home/travis/build/doctrine/DoctrineORMModule/vendor/doctrine/doctrine-module has uncommitted changes."

**Comment by devtosystems** [comment]
Referenced here #475

**Comment by Danielss89** [comment]
What is the status on this?
@Ocramius @gianarb 
I'm currently in the process of upgrade ZfcUserDoctrineORM to work with ZF3, and need DoctrineORMModule to work with zend-mvc: ~3.0

**Comment by Saeven** [comment]
Are any maintainers open to reviewing this PR?  Confirmation that this module is not abandoned would be great.

**Comment by MatthiasKuehneEllerhold** [comment]
Another PR with ZF3 from @webimpress was merged to master: #493
This PR here had a lot more changes in it that the maintainers didnt want to approve together.
The current master is compatible with ZF3 (except for #496). So yeah this module is not abandoned.

**Comment by Saeven** [comment]
Thank you very much @MatthiasKuehneEllerhold!

**Comment by webimpress** [comment]
@MatthiasKuehneEllerhold thanks. I didn't close this PR because maybe someone would like to move some parts of it into dev-master. I don't have time to prepare separate PRs with small changes, but I think it will be worth to do that...

**Comment by gianarb** [comment]
I am one of the maintainer @saeven. This module is followed and we are working to make it ready to work with zf3. I was in holiday

**Comment by samsonasik** [comment]
@webimpress it needs rebase

**Comment by Saeven** [comment]
@gianarb Thanks for your efforts!

**Comment by samsonasik** [comment]
@webimpress could you rebase ? /cc @gianarb

**Comment by webimpress** [comment]
@samsonasik I think this PR is not gonna be merged, because it drops support for ZF2. Already ZF3 support is merged to master branch - https://github.com/doctrine/DoctrineORMModule/pull/493. I left this PR open because maybe someone will move some part from it to next release.
There are still some issues with doctrine modules, probably this is why it's still not released.
As far as I know there is a problem with installing modules using zend-component-installer (I'm working on it) and also with console tools (when modules are loaded using PSR-4).

**Comment by webimpress** [comment]
No longer needed. DoctrineORMModule supports ZF2 and ZF3 since 1.1.0 release, please see: https://github.com/doctrine/DoctrineORMModule/pull/493
